feat: Lindemann-Weierstrass Theorem

astrainfinita opened the pull request:
---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

There's still a lot of work to be done before it gets into mathlib.

- [x] depends on: #6593
- [x] depends on: #6719
- [x] depends on: #6723
- [x] depends on: #6740
- [x] depends on: #7274
- [x] depends on: #8807
- [x] depends on: #12191
- [x] depends on: #12764
- [x] depends on: #16259
- [x] depends on: #16262
- [x] depends on: #16886
- [x] depends on: #16898
- [x] depends on: #16923
- [x] depends on: #16933
- [x] depends on: #17119
- [x] depends on: #17893
- [x] depends on: #17894 
- [x] depends on: #17896
- [x] depends on: #18556
- [ ] depends on: #18693
- [x] depends on: #19225 
- [x] depends on: #19246
- [x] depends on: #20092

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

Ruben-VandeVelde commented:
fyi @j-loreaux: I needed to add an instance from #5602 to fix the build here after merging master

j-loreaux commented:
That's fine. It should be back in after the update PR. I had only removed it so that the #5602 was cleanly reverted.

alreadydone commented on Mathlib/Algebra/MonoidAlgebra/Equiv.lean:
```diff
@@ -0,0 +1,143 @@
+namespace AddMonoidAlgebra
+
+@[simps]
+def ringEquivCongrLeft {R S G : Type _} [Semiring R] [Semiring S] [AddMonoid G] (f : R ≃+* S) :
+    AddMonoidAlgebra R G ≃+* AddMonoidAlgebra S G :=
```
This file should be easy to PR separately.

In general I think these `Equiv`s should be obtained by combining both directions (`RingHom`s in this case). We should show that `AddMonoidAlgebra · G` is a functor (prove the `comp` and `id` lemmas) and deduce that both directions are inverse to each other.

Some assumptions might be able to be generalized, e.g. here [AddZeroClass G] should work.

alreadydone commented on Mathlib/NumberTheory/Transcendental/LindemannWeierstrass.lean:
```diff
@@ -0,0 +1,1444 @@
+    toConjEquiv_apply_zero_eq]
+  simp_rw [toConjEquiv_symm_single, MulMemClass.mk_mul_mk]
+  haveI := Nat.noZeroSMulDivisors ℚ F
+  simp_rw [Finsupp.indicator_const_eq_sum_single, sum_mul, mul_sum,
```
```suggestion
  simp_rw [Finsupp.indicator_eq_sum_single, sum_mul, mul_sum,
```

alreadydone commented on Mathlib/Algebra/BigOperators/Finsupp.lean:
```diff
@@ -629,6 +633,11 @@ lemma prod_indicator_index [Zero M] [CommMonoid N]
+theorem prod_indicator_const_index [Zero M] [CommMonoid N] {s : Finset α} (m : M) {h : α → M → N}
+    (h_zero : ∀ a ∈ s, h a 0 = 1) : (indicator s fun _ _ => m).prod h = ∏ x in s, h x m :=
+  prod_indicator_index _ h_zero
+
```
I think these can be removed now (more general versions were merged)

alreadydone commented on Mathlib/FieldTheory/GalConj.lean:
```diff
@@ -0,0 +1,394 @@
+
+variable [IsSeparable F E]
+
+noncomputable nonrec def minpoly : GalConjClasses F E → F[X] :=
+  Quotient.lift (minpoly F) fun (a b : E) ⟨f, h⟩ =>
+    minpoly.eq_of_algHom_eq f.symm.toAlgHom f.symm.injective (IsSeparable.isIntegral F a)
+      (h ▸ (f.symm_apply_apply b).symm)
```
This seems to be the only place `minpoly.eq_of_algHom_eq` is used. Should be easy to use [minpoly.minpoly_algHom](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/Minpoly/Basic.html#minpoly.minpoly_algHom) instead?

eric-wieser commented on Mathlib/Algebra/MonoidAlgebra/Equiv.lean:
This file isn't used at all by the rest of this PR; can it just be deleted?

eric-wieser commented on Mathlib/NumberTheory/Transcendental/LindemannWeierstrass.lean:
```diff
@@ -0,0 +1,1444 @@
+variable (F : Type _) [Field F] [Algebra ℚ F]
+
+noncomputable def mapDomainFixed : Subalgebra F (AddMonoidAlgebra F (K s)) where
+  carrier x := ∀ f : Gal s, AddMonoidAlgebra.mapDomainAlgAut ℚ _ f.toAddEquiv x = x
```
```suggestion
  carrier := {x | ∀ f : Gal s, AddMonoidAlgebra.mapDomainAlgAut ℚ _ f.toAddEquiv x = x}
```
This is supposed to be a set

alreadydone commented on Mathlib/Data/Polynomial/Derivative.lean:
```diff
@@ -669,6 +680,10 @@ theorem iterate_derivative_X_sub_pow (n k : ℕ) (c : R) :
 set_option linter.uppercaseLean3 false in
 #align polynomial.iterate_derivative_X_sub_pow Polynomial.iterate_derivative_X_sub_pow
 
+theorem iterate_derivative_X_sub_pow' (n k : ℕ) (c : R) :
+    (derivative^[k]) ((X - C c) ^ n : R[X]) = n.descFactorial k • (X - C c) ^ (n - k) := by
+  rw [iterate_derivative_X_sub_pow, Nat.descFactorial_eq_prod_range, nsmul_eq_mul]
```
`iterate_derivative_X_add_pow`, `iterate_derivative_X_sub_pow`, `coeff_iterate_derivative_as_prod_range` are not used anywhere, so I guess we could just change their statements?

I haven't looked at Derivative2.lean carefully, but some stuff there look like they could be PR'd into this file ...

alreadydone commented on Mathlib/NumberTheory/Transcendental/LindemannWeierstrass.lean:
```diff
@@ -0,0 +1,1441 @@
+
+open GalConjClasses
+
+def toConjEquiv : mapDomainFixed s F ≃ (GalConjClasses ℚ (K s) →₀ F) := by
```
The LHS can be replaced by the subset of `A →₀ F` consisting of elements `x` such that `x a = x b` whenever `a ~ b` for some equivalence relation `~` on `A`, and `GalConjClasses ℚ (K s)` would be replaced by the quotient by the equivalence relation. This works for `Pi` always, but for `Finsupp` the right-to-left direction needs the equivalence classes to be finite.

The next 150+ lines seem to be mainly using this `Equiv` to transfer algebraic instances on the LHS to the RHS, and it's not yet clear to me why this is necessary; could we not just use this Equiv to map to the LHS and do all algebraic operations there?

Ruben-VandeVelde committed:
Reduce variants about polynomial derivatives

Ruben-VandeVelde committed:
Reduce diff

alreadydone commented on Mathlib/NumberTheory/Transcendental/LindemannWeierstrass.lean:
```diff
@@ -0,0 +1,1441 @@
+
+section
+
+variable (s : Finset ℂ)
+
+namespace Transcendental -- Conflict with Mathlib.NumberTheory.Dioph
+abbrev Poly : ℚ[X] :=
+  ∏ x in s, minpoly ℚ x
```
I'm wondering whether i would be easier to take the [normalClosure](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/NormalClosure.html#normalClosure) of `IntermediateField.adjoin ℚ s` in `ℂ`, or we can do it in one step by adjoining all conjugates (same minpoly) of elements in `s`. Or maybe we can directly work with automorphisms of ℂ rather than of an intermediate field?

alreadydone committed:
minor

Ruben-VandeVelde committed:
Update Mathlib/NumberTheory/Transcendental/LindemannWeierstrass.lean

Co-authored-by: Junyan Xu <junyanxu.math@gmail.com>

Ruben-VandeVelde committed:
Speed up

Ruben-VandeVelde committed:
Add shortcuts

Ruben-VandeVelde committed:
Merge remote-tracking branch 'origin/master' into transcendental

Ruben-VandeVelde committed:
Merge remote-tracking branch 'origin/master' into FundThm_SymPoly

Ruben-VandeVelde committed:
Fix

Ruben-VandeVelde committed:
Merge branch 'FundThm_SymPoly' into transcendental

Ruben-VandeVelde committed:
Update Mathlib/RingTheory/MvPolynomial/Symmetric/Eval.lean

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Ruben-VandeVelde committed:
Fix

Ruben-VandeVelde committed:
Fix

Ruben-VandeVelde committed:
Merge branch 'master' into transcendental

Ruben-VandeVelde committed:
fix

Ruben-VandeVelde committed:
partial fix

Ruben-VandeVelde committed:
fix

Ruben-VandeVelde committed:
fix

Ruben-VandeVelde committed:
fix

astrainfinita committed:
revert commit d46a030950662204744f2c467572c6baa5a9e195 aroot notation

astrainfinita committed:
Merge branch 'master' into transcendental

astrainfinita committed:
fix & `Type _` -> `Type*`

astrainfinita committed:
fix & golf

astrainfinita committed:
fix & golf

astrainfinita committed:
remove heq lemmas

astrainfinita committed:
namespace

astrainfinita committed:
remove some `AddMonoidAlgebra` apis

astrainfinita committed:
equiv

astrainfinita committed:
remove warning

astrainfinita committed:
still need this...

astrainfinita committed:
instance & style

astrainfinita committed:
fix

astrainfinita committed:
fix

astrainfinita committed:
fix

astrainfinita committed:
Merge branch 'master' into transcendental

astrainfinita committed:
move

astrainfinita committed:
fix

astrainfinita committed:
split

astrainfinita committed:
golf

astrainfinita committed:
golf

astrainfinita committed:
move & split

astrainfinita committed:
golf

astrainfinita committed:
chore(FieldTheory/PolynomialGaloisGroup): move lemmas, reduce imports

#dependencies:
`Mathlib.FieldTheory.PolynomialGaloisGroup`: 1826->1323
`Mathlib.Analysis.Complex.Polynomial`: 1811->1826

I'm not sure if we should add a new file.

astrainfinita committed:
reduce imports

astrainfinita committed:
fix

astrainfinita committed:
Merge branch 'FR_split_polynomialGaloisGroup' into transcendental

Ruben-VandeVelde committed:
Merge remote-tracking branch 'origin/master' into supDegree_LinearOrder

Ruben-VandeVelde committed:
Review

Ruben-VandeVelde committed:
Merge branch 'supDegree_LinearOrder' into FundThm_SymPoly

Ruben-VandeVelde committed:
Fix

Ruben-VandeVelde committed:
shake

Ruben-VandeVelde committed:
Merge remote-tracking branch 'origin/master' into supDegree_LinearOrder

Ruben-VandeVelde committed:
Merge branch 'supDegree_LinearOrder' into FundThm_SymPoly

Ruben-VandeVelde committed:
Merge branch 'FundThm_SymPoly' into transcendental

Ruben-VandeVelde committed:
Mathlib.lean

Ruben-VandeVelde committed:
Overlong line

Ruben-VandeVelde committed:
fix

Ruben-VandeVelde committed:
fix

Ruben-VandeVelde committed:
fix

Ruben-VandeVelde committed:
fix

Ruben-VandeVelde committed:
shake autofix

Ruben-VandeVelde committed:
Merge remote-tracking branch 'origin/master' into supDegree_LinearOrder

Ruben-VandeVelde committed:
Merge branch 'supDegree_LinearOrder' into FundThm_SymPoly

Ruben-VandeVelde committed:
Merge branch 'FundThm_SymPoly' into transcendental

Ruben-VandeVelde committed:
unused argument

Ruben-VandeVelde committed:
def/lemma, cleanup

Ruben-VandeVelde committed:
Merge remote-tracking branch 'origin/master' into supDegree_LinearOrder

Ruben-VandeVelde committed:
Merge branch 'supDegree_LinearOrder' into FundThm_SymPoly

Ruben-VandeVelde committed:
Merge branch 'FundThm_SymPoly' into transcendental

Ruben-VandeVelde committed:
Merge remote-tracking branch 'origin/master' into supDegree_LinearOrder

Ruben-VandeVelde committed:
Merge branch 'master' into supDegree_LinearOrder

Ruben-VandeVelde committed:
fix

Ruben-VandeVelde committed:
Merge branch 'supDegree_LinearOrder' into FundThm_SymPoly

Ruben-VandeVelde committed:
Merge branch 'FundThm_SymPoly' into transcendental

Ruben-VandeVelde committed:
fix

Ruben-VandeVelde committed:
fix

Ruben-VandeVelde committed:
fix

Ruben-VandeVelde committed:
fix

Ruben-VandeVelde committed:
Reuse landed code

Ruben-VandeVelde committed:
Merge remote-tracking branch 'origin/master' into supDegree_LinearOrder

Ruben-VandeVelde committed:
Merge branch 'supDegree_LinearOrder' into FundThm_SymPoly

Ruben-VandeVelde committed:
Merge branch 'FundThm_SymPoly' into transcendental

Ruben-VandeVelde committed:
fix

Ruben-VandeVelde committed:
warning

Ruben-VandeVelde committed:
Merge branch 'master' into supDegree_LinearOrder

Ruben-VandeVelde committed:
Merge branch 'supDegree_LinearOrder' into FundThm_SymPoly

Ruben-VandeVelde committed:
Merge branch 'FundThm_SymPoly' into transcendental

Ruben-VandeVelde committed:
Merge branch 'master' into supDegree_LinearOrder

Ruben-VandeVelde committed:
Merge branch 'supDegree_LinearOrder' into FundThm_SymPoly

Ruben-VandeVelde committed:
Merge branch 'FundThm_SymPoly' into transcendental

Ruben-VandeVelde committed:
Merge branch 'master' into supDegree_LinearOrder

Ruben-VandeVelde committed:
Merge branch 'supDegree_LinearOrder' into transcendental

Ruben-VandeVelde committed:
Merge branch 'supDegree_LinearOrder' into FundThm_SymPoly

Ruben-VandeVelde committed:
Merge branch 'master' into supDegree_LinearOrder

Ruben-VandeVelde committed:
Merge branch 'supDegree_LinearOrder' into FundThm_SymPoly

Ruben-VandeVelde committed:
Merge branch 'FundThm_SymPoly' into transcendental

Ruben-VandeVelde committed:
fix

Ruben-VandeVelde committed:
Merge remote-tracking branch 'origin/master' into supDegree_LinearOrder

Ruben-VandeVelde committed:
Merge branch 'supDegree_LinearOrder' into FundThm_SymPoly

Ruben-VandeVelde committed:
Merge branch 'FundThm_SymPoly' into transcendental

Ruben-VandeVelde committed:
up

Ruben-VandeVelde committed:
up

Ruben-VandeVelde committed:
up

Ruben-VandeVelde committed:
Merge branch 'master' into supDegree_LinearOrder

Ruben-VandeVelde committed:
fix

Ruben-VandeVelde committed:
Merge branch 'supDegree_LinearOrder' into FundThm_SymPoly

Ruben-VandeVelde committed:
fix

Ruben-VandeVelde committed:
Merge branch 'FundThm_SymPoly' into transcendental

Ruben-VandeVelde committed:
fix

Ruben-VandeVelde committed:
fix

Ruben-VandeVelde committed:
fix

Ruben-VandeVelde committed:
Limit to LinearOrder

Ruben-VandeVelde committed:
tidy

Parcly-Taxel committed:
Merge branch 'master' into supDegree_LinearOrder

Parcly-Taxel committed:
Update Mathlib/Algebra/Group/Subsemigroup/Operations.lean

Co-authored-by: Yaël Dillies <yael.dillies@gmail.com>

Parcly-Taxel committed:
Merge branch 'supDegree_LinearOrder' of github.com:leanprover-community/mathlib4 into supDegree_LinearOrder

Ruben-VandeVelde committed:
Golf

This does not appear to slow down the proof noticeably.

Ruben-VandeVelde committed:
Merge remote-tracking branch 'origin/supDegree_LinearOrder' into FundThm_SymPoly

Ruben-VandeVelde committed:
Merge branch 'FundThm_SymPoly' into transcendental

Ruben-VandeVelde committed:
Merge remote-tracking branch 'origin/master' into supDegree_LinearOrder

Ruben-VandeVelde committed:
review

Ruben-VandeVelde committed:
Apply suggestions from code review

Co-authored-by: Yaël Dillies <yael.dillies@gmail.com>

Ruben-VandeVelde committed:
fix

Ruben-VandeVelde committed:
Merge branch 'supDegree_LinearOrder' into FundThm_SymPoly

Ruben-VandeVelde committed:
fix

Ruben-VandeVelde committed:
Merge branch 'FundThm_SymPoly' into transcendental

Ruben-VandeVelde committed:
fix

Ruben-VandeVelde committed:
tidy

Ruben-VandeVelde committed:
fix

Parcly-Taxel committed:
Merge branch 'master' into FundThm_SymPoly

Ruben-VandeVelde committed:
fix

Ruben-VandeVelde committed:
tidy and review names

Ruben-VandeVelde committed:
fix

Ruben-VandeVelde committed:
Merge branch 'FundThm_SymPoly' into transcendental

Ruben-VandeVelde committed:
fix

Ruben-VandeVelde committed:
Merge branch 'master' into transcendental

Ruben-VandeVelde committed:
fix

Ruben-VandeVelde committed:
fix

Ruben-VandeVelde committed:
feat: sum of iterated derivatives

Co-authored-by: negiizhao <egresf@gmail.com>

Ruben-VandeVelde committed:
Merge branch 'SumIteratedDerivative' into transcendental

Ruben-VandeVelde committed:
simp

Ruben-VandeVelde committed:
fix

Ruben-VandeVelde committed:
Merge remote-tracking branch 'origin/master' into transcendental

Ruben-VandeVelde committed:
fix

Ruben-VandeVelde committed:
review

astrainfinita committed:
cleanup

jjdishere committed:
Create IsConjRoot.lean

Ruben-VandeVelde committed:
Update SumIteratedDerivative.lean

astrainfinita committed:
Merge branch 'SumIteratedDerivative' into transcendental

astrainfinita committed:
`derivativeFinsupp`

astrainfinita committed:
move

astrainfinita committed:
do not need `FloorRing.exists_prime_mul_pow_div_factorial_lt_one`

astrainfinita committed:
`derivativeFinsupp` lemmas

astrainfinita committed:
feat: `FloorSemiring.eventually_mul_pow_div_factorial_lt`

astrainfinita committed:
Update Prime.lean

astrainfinita committed:
Update Prime.lean

astrainfinita committed:
Update Basic.lean

astrainfinita committed:
Merge branch 'master' into transcendental

astrainfinita committed:
Merge branch 'SumIteratedDerivative' into transcendental

astrainfinita committed:
Merge branch 'FR_eventually_mul_pow_div_factorial_lt' into transcendental

astrainfinita committed:
fix

astrainfinita committed:
suggestion

Ruben-VandeVelde committed:
address review comments and generalize

astrainfinita committed:
cleanup

jjdishere committed:
Update IsConjRoot.lean

Ruben-VandeVelde committed:
Merge branch 'master' into SumIteratedDerivative

Ruben-VandeVelde commented:
I'll fix the merge conflicts tomorrow.

Ruben-VandeVelde committed:
fix

Ruben-VandeVelde committed:
Merge branch 'SumIteratedDerivative' into transcendental

Ruben-VandeVelde committed:
fix

astrainfinita committed:
fix

astrainfinita committed:
cleanup

astrainfinita committed:
protect minpoly

astrainfinita committed:
maybe a better definition

needs to be cleaned up. see also https://github.com/leanprover-community/mathlib4/pull/6718#discussion_r1328899532

jjdishere committed:
Update IsConjRoot.lean

jjdishere committed:
Merge branch 'master' into jiedong_jiang_is_conj_root

jjdishere committed:
Update IsConjRoot.lean

jjdishere committed:
Update IsConjRoot.lean

jjdishere committed:
Update IsConjRoot.lean

jjdishere committed:
no sorry

jjdishere committed:
Merge branch 'master' into jiedong_jiang_is_conj_root

jjdishere committed:
docs

jjdishere committed:
Merge branch 'master' into jiedong_jiang_is_conj_root

jjdishere committed:
Update IsConjRoot.lean

jjdishere committed:
Update IsConjRoot.lean

jjdishere committed:
docs

jjdishere committed:
lake exe mk_all

jjdishere committed:
Update Mathlib/FieldTheory/Minpoly/IsConjRoot.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

jjdishere committed:
align docs

jjdishere committed:
change x' to y

Ruben-VandeVelde committed:
Merge branch 'master' into transcendental

Ruben-VandeVelde committed:
fix

Ruben-VandeVelde committed:
fix

Ruben-VandeVelde committed:
fix

Ruben-VandeVelde committed:
lint

Ruben-VandeVelde committed:
tidy

Ruben-VandeVelde committed:
fixes

Ruben-VandeVelde committed:
Merge branch 'master' into transcendental

Ruben-VandeVelde committed:
lint

alreadydone commented on Mathlib/FieldTheory/GalConj.lean:
```diff
@@ -0,0 +1,323 @@
+    simp [NormalClosure.algebraAux_def, AlgEquiv.eq_symm_apply]
+
+def IsGalConj.setoid :=
+  (MulAction.orbitRel (NormalClosure F E ≃ₐ[F] NormalClosure F E) (NormalClosure F E)).comap
```
This is quite a restrictive definition (only applies to NormalClosure). I recall it was defined using `minpoly` before; am I hallucinating? If #17783 is merged this file should probably switch to use it?

jjdishere replied:
I‘ve discussed with @FR-vdash-bot and we believe that the definition here and in #17783 is indeed mathematically equivalent, even in the case of transcendental extensions. (Although some more api's need to be built.) We agreed to use the `minpoly K x = minpoly K y` definition since it involves less.

astrainfinita committed:
Merge remote-tracking branch 'origin/jiedong_jiang_is_conj_root' into FR_conjrootclass

astrainfinita committed:
feat: define `ConjRootClass` and prove its basic properties

astrainfinita committed:
Merge branch 'FR_conjrootclass' into transcendental

astrainfinita committed:
cleanup

astrainfinita committed:
cleanup

astrainfinita committed:
cleanup

astrainfinita committed:
mk_all

astrainfinita committed:
Merge branch 'master' into transcendental

astrainfinita committed:
Merge branch 'master' into transcendental

astrainfinita committed:
fix

astrainfinita committed:
cleanup, generalize

astrainfinita committed:
Merge branch 'master' into transcendental

astrainfinita committed:
cleanup

astrainfinita committed:
cleanup

astrainfinita committed:
do not compute on `ConjRootClass K G →₀ L`

Ruben-VandeVelde committed:
Merge remote-tracking branch 'origin/master' into transcendental

astrainfinita committed:
cleanup

astrainfinita committed:
fix

astrainfinita committed:
cleanup

astrainfinita committed:
cleanup, do not need `Eval` now

astrainfinita committed:
cleanup, do not need global aux defs now

astrainfinita committed:
shake, fix

astrainfinita committed:
cleanup

astrainfinita committed:
cleanup

astrainfinita committed:
reference

astrainfinita committed:
cleanup & algebraicIndependent_exp

alreadydone commented on Mathlib/NumberTheory/Transcendental/Lindemann/Basic.lean:
```diff
@@ -205,8 +206,20 @@ private theorem linearIndependent_exp' (u : ι → ℂ) (hu : ∀ i, IsIntegral
+      (v ·) (v · |>.2) (by simpa [sum_attach _ fun x ↦ v x * cexp (u x)])
+
+theorem algebraicIndependent_exp (u : ι → integralClosure ℚ ℂ) (hu : LinearIndependent ℕ u) :
+    AlgebraicIndependent (integralClosure ℚ ℂ) fun i ↦ exp (u i) := by
```
AlgebraicIndependent over `integralClosure ℚ ℂ` is the same as AlgebraicIndependent over ℚ (or ℤ), right? Do we have this? @acmepjz

acmepjz replied:
Seems that we don't have this yet. Let me check later.

alreadydone replied:
> Looks like [Algebra.IsAlgebraic.trans](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/Algebraic.html#Algebra.IsAlgebraic.trans) and [IntermediateField.isAlgebraic_adjoin_iff_isAlgebraic](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/AlgebraicClosure.html#IntermediateField.isAlgebraic_adjoin_iff_isAlgebraic) are two things to be generalized.

I've generalized these to domains and proven the Transcendental version of this [here](https://github.com/leanprover-community/mathlib4/compare/Subalgebra.algebraicClosure?expand=1#diff-1fed4ce28689dbfee7313dcc400e1b532d2bc3e97b932d74791f8506e07f9083R932-R939) but I'll wait for #6183 and #19370 before making a PR.

alreadydone replied:
> By the way, [Transcendental.of_tower_top_of_injective](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/Algebraic.html#Transcendental.of_tower_top_of_injective) should be renamed to `Transcendental.restrictScalars` and [IsAlgebraic.tower_top_of_injective](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/Algebraic.html#IsAlgebraic.tower_top_of_injective) should be renamed to `IsAlgebraic.extendScalars`, as someone already pointed out that the original naming is not appropriate.

Why were `IsAlgebraic.tower_top_of_subalgebra_le` and `Transcendental.of_tower_top_of_subalgebra_le` not changed to `extend/restrictScalars`?

alreadydone replied:
I've proven [`AlgebraicIndependent.extendScalars`](https://github.com/leanprover-community/mathlib4/compare/Subalgebra.algebraicClosure?expand=1#diff-79fb72674b5880e7c904ddfcdbb728fd400ab58285267410bccebb4b118337baR627-R628); will merge master and clean up when I get up.

Ruben-VandeVelde replied:
@alreadydone I don't know what you want to change here - feel free to push directly to this PR

alreadydone replied:
Now it suffices to show `AlgebraicIndependent ℚ fun i ↦ exp (u i)` but that may not make the proof easier ...

Ruben-VandeVelde committed:
Merge branch 'master' into transcendental

jcommelin committed:
golfs in the analytical part

jcommelin commented on Mathlib/NumberTheory/Transcendental/Lindemann/Init/AnalyticalPart.lean:
```diff
@@ -0,0 +1,185 @@
+import Mathlib.Topology.Algebra.Polynomial
+
+/-!
+# The Lindemann-Weierstrass theorem
```
My impression is that this file is pretty close to being ready for merging.
It would of course need docstrings and some more comments and documentation.

Ruben-VandeVelde committed:
log + tidy

Ruben-VandeVelde committed:
clarify code

Ruben-VandeVelde committed:
unused args

Ruben-VandeVelde committed:
simplify

Ruben-VandeVelde committed:
Merge remote-tracking branch 'origin/master' into transcendental

Ruben-VandeVelde committed:
fix

Ruben-VandeVelde committed:
fix

Ruben-VandeVelde committed:
golf

Ruben-VandeVelde committed:
lint

Ruben-VandeVelde committed:
Reorganize a bit

Ruben-VandeVelde committed:
wip

Ruben-VandeVelde committed:
Merge branch 'master' into transcendental

Ruben-VandeVelde committed:
fix bad merge

Ruben-VandeVelde committed:
More clarifications for the analytic part

Ruben-VandeVelde committed:
typos

Ruben-VandeVelde committed:
Merge branch 'master' into transcendental

jcommelin commented on Mathlib/NumberTheory/Transcendental/Lindemann/Init/AnalyticalPart.lean:
```diff
@@ -45,9 +47,15 @@ theorem integral_exp_mul_eval (p : ℂ[X]) (s : ℂ) :
 
 /--
-`P` is a slightly generalized version of `Iᵢ` in
+In what follows, we will use integration by parts to rewrite provide an integral form for `P`
```
Upstreaming the diff in this file seems like a natural next PR to me.

Ruben-VandeVelde committed:
fix

Ruben-VandeVelde committed:
warning

Ruben-VandeVelde committed:
warning

Ruben-VandeVelde committed:
Merge branch 'master' into transcendental

Ruben-VandeVelde committed:
fix

Ruben-VandeVelde committed:
fix

Ruben-VandeVelde committed:
fix

Ruben-VandeVelde committed:
fix

astrainfinita commented:
This PR has been migrated to a fork-based workflow: https://github.com/leanprover-community/mathlib4/pull/28013
